README.md
# Memoization

A general introduction into memoization: https://en.wikipedia.org/wiki/Memoization

**Memoization** is a simple npm package to cache the result of the provided function for a limited amount of time.

The purpose of this project is to demonstrate the followings:

- Development of a simple npm package with javascript.
- Automatic Unit Tests and achieving 100% coverage.
- Using CI/CD tools:
  - [Github Actions](https://github.com/features/actions) to automatically run tests and publish the npm package.
- Using Automatic Versioning:
  - Use [Semantic Release](https://github.com/semantic-release/semantic-release) to calculate new version number by processing commit messages.
  - Publish the `next` branch to the `next` channel on npm and publish the `master` branch to the main channel.

## Instalation

```
yarn add @emech/memoization
```

## Usage

```javascript
const memoize = require("@emech/memoization").memoize;

const random = (max) => Math.random() * max;
const memoizedRandom = memoize(random, (max) => max, 1000);

const randomNumber = memoizedRandom(2);
const randomNumber2 = memoizedRandom(2);

assert(randomNumber === randomNumber2);
```

## Api
```javascript
memoize(realFunc [, resolver], timout) => memoizedFunc
```
- **realFunc**: `function` The function for which the return values should be cached.
- **resolver**: `function` *(Optional)* If provided gets called for each function call with the exact same set of parameters as the original function, the resolver function should provide the memoization key.
- **timeout**: `number` Timeout for cached values in milliseconds


## Contribution

```bash
# Clone
$ git clone git@github.com:emech-en/memoization.git

# Run test 
$ yarn test

# Report coverage
$ yarn test:coverage
```

## Stay in touch

- Author: [Emech En](https://github.com/emech-en) < [mh.niroomand91@gmail.com](mailto://mh.niroomand91@gmail.com) >


CHANGELOG.md
## [1.2.2](https://github.com/emech-en/memoization/compare/v1.2.1...v1.2.2) (2020-06-16)


### Bug Fixes

* update README.md (For updating npm page) ([ded7a6b](https://github.com/emech-en/memoization/commit/ded7a6b045b7caec12c1b578fa6256f44cfddde9))

## [1.2.1](https://github.com/emech-en/memoization/compare/v1.2.0...v1.2.1) (2020-06-16)


### Bug Fixes

* update README.md ([5a4ae4f](https://github.com/emech-en/memoization/commit/5a4ae4f01273cbfe9b3d7d27ab057fa8f47ff118))

# [1.2.0](https://github.com/emech-en/memoization/compare/v1.1.0...v1.2.0) (2020-06-16)


### Bug Fixes

* update README.md on npmjs.com ([7c54084](https://github.com/emech-en/memoization/commit/7c54084de2857cf94505b28db85293fd5cb9d112))


### Features

* support non-premitiv keys ([#2](https://github.com/emech-en/memoization/issues/2)) ([86a37c0](https://github.com/emech-en/memoization/commit/86a37c0f3c7fae4a9b7d57002269d97de3a011df))

## [1.1.1](https://github.com/emech-en/memoization/compare/v1.1.0...v1.1.1) (2020-06-11)


### Bug Fixes

* update README.md on npmjs.com ([7c54084](https://github.com/emech-en/memoization/commit/7c54084de2857cf94505b28db85293fd5cb9d112))

# [1.1.0](https://github.com/emech-en/memoization/compare/v1.0.1...v1.1.0) (2020-06-11)


### Features

* consider default resolver ([fc9a0fb](https://github.com/emech-en/memoization/commit/fc9a0fbf93630d9382c0c61a83455f0c9bebf17b))

## [1.0.1](https://github.com/emech-en/memoization/compare/v1.0.0...v1.0.1) (2020-06-11)


### Bug Fixes

* run tests before publish ([865300e](https://github.com/emech-en/memoization/commit/865300ea385b0d64a34f2ae4216f0f40c9d47f76))

# 1.0.0 (2020-06-11)


### Features

* implement simple memoize function ([d4f457c](https://github.com/emech-en/memoization/commit/d4f457ce8762ef4473c8b9bba4231488efe56268))


test.js
const memoization = require("./memoizaton");
const expect = require("chai").expect;
const sinon = require("sinon");
// hint: use https://sinonjs.org/releases/v6.1.5/fake-timers/ for faking timeouts

describe("memoization", function () {
  let clock;
  before(() => {
    clock = sinon.useFakeTimers(Date.now());
  });
  after(() => {
    clock.uninstall();
  });

  it("should memoize function result", () => {
    const testFunction = (year, month, day) => {
      return Date.now() + Date(year, month, day);
    };

    const testMemoizedFunction = memoization.memoize(
      testFunction,
      (year, month, day) => year + month + day,
      1000
    );

    const firstResult = testMemoizedFunction(1, 11, 26);
    const secondResult = testMemoizedFunction(1, 11, 26);
    expect(firstResult).to.equal(secondResult);

    clock.tick(500);
    const thirdResult = testMemoizedFunction(1, 11, 26);
    expect(thirdResult).to.equal(secondResult);

    clock.tick(1000);
    const forthResult = testMemoizedFunction(1, 11, 26);
    expect(forthResult).not.to.equal(secondResult);
  });
});


memoizaton.test.js
const memoization = require("./memoizaton");
const sinon = require("sinon");
// hint: use https://sinonjs.org/releases/v6.1.5/fake-timers/ for faking timeouts

describe("memoization", () => {
  let clock;
  beforeEach(() => {
    clock = sinon.useFakeTimers(Date.now());
  });
  afterEach(() => {
    clock.uninstall();
  });

  it("should memoize function result", () => {
    const testFunction = (year, month, day) => {
      return Date.now() + Date(year, month, day);
    };

    const testMemoizedFunction = memoization.memoize(
      testFunction,
      (year, month, day) => year + month + day,
      1000
    );

    const firstResult = testMemoizedFunction(1, 11, 26);
    const secondResult = testMemoizedFunction(1, 11, 26);
    expect(firstResult).toBe(secondResult);

    clock.tick(500);
    const thirdResult = testMemoizedFunction(1, 11, 26);
    expect(thirdResult).toBe(secondResult);

    clock.tick(1000);
    const forthResult = testMemoizedFunction(1, 11, 26);
    expect(forthResult).not.toBe(secondResult);
  });
});


memoization.test.js
const memoization = require("./memoization");
const sinon = require("sinon");
// hint: use https://sinonjs.org/releases/v6.1.5/fake-timers/ for faking timeouts

describe("memoization", () => {
  let clock;
  beforeEach(() => {
    clock = sinon.useFakeTimers(Date.now());
  });
  afterEach(() => {
    clock.uninstall();
  });

  it("should memoize function result for a definite time", () => {
    const testFunction = (year, month, day) => {
      return Date.now() + Date(year, month, day);
    };

    const testMemoizedFunction = memoization.memoize(
      testFunction,
      (year, month, day) => year + month + day,
      1000
    );

    const firstResult = testMemoizedFunction(1, 11, 26);
    clock.tick(100);
    const secondResult = testMemoizedFunction(1, 11, 26);
    expect(firstResult).toBe(secondResult);
    const anotherResult = testMemoizedFunction(2, 12, 26);
    expect(anotherResult).not.toBe(firstResult);

    clock.tick(500);
    const thirdResult = testMemoizedFunction(1, 11, 26);
    expect(thirdResult).toBe(secondResult);

    clock.tick(1000);
    const forthResult = testMemoizedFunction(1, 11, 26);
    expect(forthResult).not.toBe(secondResult);
  });

  it("should use default resolver: memoize(func, timeout)", () => {
    const testFunction = (year, month, day) => {
      return Date.now() + Date(year, month, day);
    };

    const testMemoizedFunction = memoization.memoize(testFunction, 1000);

    const firstResult = testMemoizedFunction(1, 11, 26);
    clock.tick(100);
    const secondResult = testMemoizedFunction(1, 11, 26);
    expect(firstResult).toBe(secondResult);
    const anotherResult = testMemoizedFunction(2, 12, 26);
    expect(anotherResult).not.toBe(firstResult);
  });

  it("should throw error if params are not valid", () => {
    const testFunction = (year, month, day) => {
      return Date.now() + Date(year, month, day);
    };

    expect(() => memoization.memoize(testFunction)).toThrow();
    expect(() => memoization.memoize(1, (key) => key, 1000)).toThrow();
    expect(() => memoization.memoize(testFunction, "hello")).toThrow();
  });

  it("should handle undefined as cache key", () => {
    const testFunction = () => {
      return Math.random();
    };

    const testMemoizedFunction = memoization.memoize(testFunction, 1000);

    const firstResult = testMemoizedFunction();
    clock.tick(100);
    const secondResult = testMemoizedFunction();
    expect(firstResult).toBe(secondResult);

    clock.tick(1000);
    const forthResult = testMemoizedFunction();
    expect(forthResult).not.toBe(secondResult);
  });

  it("should handle null as cache key", () => {
    const testFunction = () => {
      return Math.random();
    };

    const testMemoizedFunction = memoization.memoize(
      testFunction,
      () => null,
      1000
    );

    const firstResult = testMemoizedFunction();
    clock.tick(100);
    const secondResult = testMemoizedFunction();
    expect(firstResult).toBe(secondResult);

    clock.tick(1000);
    const forthResult = testMemoizedFunction();
    expect(forthResult).not.toBe(secondResult);
  });

  it("should cache async functions result", async () => {
    clock.restore();

    const asyncRandom = () => {
      return new Promise((resolve) =>
        setTimeout(() => resolve(Math.random()), 500)
      );
    };

    const testMemoizedFunction = memoization.memoize(
      asyncRandom,
      () => null,
      2000
    );

    let startTime = Date.now();
    const firstResult = await testMemoizedFunction();
    let duration = Date.now() - startTime;
    expect(duration).toBeGreaterThanOrEqual(500);

    startTime = Date.now();
    const secondResult = await testMemoizedFunction();
    duration = Date.now() - startTime;
    expect(duration).toBeLessThan(500);
    expect(firstResult).toBe(secondResult);

    await new Promise((resolve) => setTimeout(resolve, 2000));
    startTime = Date.now();
    const thirdResult = await testMemoizedFunction();
    duration = Date.now() - startTime;
    expect(duration).toBeGreaterThanOrEqual(500);
    expect(firstResult).not.toBe(thirdResult);
  });

  it("should handle object-literal as cache key", async () => {
    const testFunction = (year, month, day) => {
      return Date.now() + Date(year, month, day);
    };

    const testMemoizedFunction = memoization.memoize(
      testFunction,
      (year, month, day) => ({
        year,
        month,
        day,
      }),
      1000
    );

    const firstResult = testMemoizedFunction(1, 11, 26);
    clock.tick(100);
    const secondResult = testMemoizedFunction(1, 11, 26);
    expect(firstResult).toBe(secondResult);
    const anotherResult = testMemoizedFunction(2, 12, 26);
    expect(anotherResult).not.toBe(firstResult);

    clock.tick(500);
    const thirdResult = testMemoizedFunction(1, 11, 26);
    expect(thirdResult).toBe(secondResult);

    clock.tick(1000);
    const forthResult = testMemoizedFunction(1, 11, 26);
    expect(forthResult).not.toBe(secondResult);
  });

  it("should handle function as cache key", async () => {
    const testFunction = (addFunc) => {
      return addFunc(Date.now());
    };

    const testMemoizedFunction = memoization.memoize(testFunction, 1000);

    const someAddFunc = (num) => num + 5;

    const firstResult = testMemoizedFunction(someAddFunc);
    clock.tick(100);
    const secondResult = testMemoizedFunction(someAddFunc);
    expect(firstResult).toBe(secondResult);
    
    const anotherResult = testMemoizedFunction((num) => someAddFunc(num));
    expect(anotherResult).not.toBe(firstResult);

    clock.tick(500);
    const thirdResult = testMemoizedFunction(someAddFunc);
    expect(thirdResult).toBe(secondResult);

    clock.tick(1000);
    const forthResult = testMemoizedFunction(someAddFunc);
    expect(forthResult).not.toBe(secondResult);
  });
});


memoizaton.js
/**
 * Creates a function that memoizes the result of func. If resolver is provided,
 * it determines the cache key for storing the result based on the arguments provided to the memorized function.
 * By default, the first argument provided to the memorized function is used as the map cache key. The memorized values
 * timeout after the timeout exceeds. The timeout is in defined in milliseconds.
 *
 * Example:
 * function addToTime(year, month, day) {
 *  return Date.now() + Date(year, month, day);
 * }
 *
 * const memoized = memoization.memoize(addToTime, (year, month, day) => year + month + day, 5000)
 *
 * // call the provided function cache the result and return the value
 * const result = memoized(1, 11, 26); // result = 1534252012350
 *
 * // because there was no timeout this call should return the memorized value from the first call
 * const secondResult = memoized(1, 11, 26); // secondResult = 1534252012350
 *
 * // after 5000 ms the value is not valid anymore and the original function should be called again
 * const thirdResult = memoized(1, 11, 26); // thirdResult = 1534252159271
 *
 * @param func      the function for which the return values should be cached
 * @param resolver  if provided gets called for each function call with the exact same set of parameters as the
 *                  original function, the resolver function should provide the memoization key.
 * @param timeout   timeout for cached values in milliseconds
 */
function memoize(func, resolver, timeout) {
  // TODO: consider edge cases and validate params
  
  const cache = {};
  return (...args) => {
    const key = resolver(...args);

    if (!cache[key] || Date.now() > cache[key].expiresAt) {
      cache[key] = {
        expiresAt: Date.now() + timeout,
        value: func(...args),
      };
    }

    return cache[key].value;
  };
}

module.exports = {
  memoize,
};


memoization.js
/**
 * Creates a function that memoizes the result of func. If resolver is provided,
 * it determines the cache key for storing the result based on the arguments provided to the memorized function.
 * By default, the first argument provided to the memorized function is used as the map cache key. The memorized values
 * timeout after the timeout exceeds. The timeout is in defined in milliseconds.
 *
 * Example:
 * function addToTime(year, month, day) {
 *  return Date.now() + Date(year, month, day);
 * }
 *
 * const memoized = memoization.memoize(addToTime, (year, month, day) => year + month + day, 5000)
 *
 * // call the provided function cache the result and return the value
 * const result = memoized(1, 11, 26); // result = 1534252012350
 *
 * // because there was no timeout this call should return the memorized value from the first call
 * const secondResult = memoized(1, 11, 26); // secondResult = 1534252012350
 *
 * // after 5000 ms the value is not valid anymore and the original function should be called again
 * const thirdResult = memoized(1, 11, 26); // thirdResult = 1534252159271
 *
 * @param func      the function for which the return values should be cached
 * @param resolver  if provided gets called for each function call with the exact same set of parameters as the
 *                  original function, the resolver function should provide the memoization key.
 * @param timeout   timeout for cached values in milliseconds
 */
function memoize(func, resolver, timeout) {
  /**
   * Considering default resolver memoize(func, timout) { }
   */
  if (typeof resolver === "number") {
    timeout = resolver;
    resolver = (...args) => args[0];
  }

  /**
   * Check arguments types to be valid
   */
  if (
    typeof func !== "function" ||
    typeof resolver !== "function" ||
    typeof timeout !== "number"
  ) {
    throw new Error("Argument Error: memoize(function, [function,] number)");
  }

  /**
   * Use map for better performance also support non-primitiv keys
   */
  const cache = new Map();

  /**
   * Return memoized function
   */
  return (...args) => {
    let key = resolver(...args);

    /**
     *  Serialize objects keys
     */
    if (typeof key === "object") {
      key = JSON.stringify(key);
    }

    /**
     * Check also expire time for more accuracy
     */
    if (!cache.has(key) || Date.now() > cache.get(key).expiresAt) {
      cache.set(key, {
        expiresAt: Date.now() + timeout,
        value: func(...args),
        /**
         * remove cached value after timeout for memory consideration
         */
        timeout: setTimeout(() => cache.delete(key), timeout),
      });
    }

    return cache.get(key).value;
  };
}

module.exports = {
  memoize,
};
